### PR TITLE
Fix vanilla map pixel data being used in various places

### PR DESCRIPTION
**Pull request by williewillus**
#3952 for 1.12

### CODE REVIEW

**Comment by covers1624** [comment]
Yes, Oh god yes. This solves all my issues. +40 :D

**Comment by mezz** [comment]
This looks good, but can you make a test mod?
There are still some places where `== Items.FILLED_MAP` is used and I want to be sure this PR works as intended.

**Comment by mezz** [comment]
@williewillus any update?

**Comment by williewillus** [comment]
Been busy for the past week, will get one made soon ^tm and the 1.11 one updated

**Comment by williewillus** [comment]
Added a test mod.
How to use:
1. Open a superflat world, and open a vanilla filled map (see a green field)
2. Put it on a frame
3. `/give @p mapdatatest:custom_map`
4. It should render as a solid chunk of red, both in hand an on a frame
5. Without these patches, the custom map would render the same data as the vanilla map (green field)

File: src/test/java/net/minecraftforge/debug/MapDataTest.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    public void registerItems(RegistryEvent.Register<Item> evt)
+    {
+        evt.getRegistry().register(new CustomMap().setRegistryName("mapdatatest", "custom_map"));
+    }
+
+    public static class CustomMap extends ItemMap {
+        private static final String PREFIX = "custommap";
+
+        // copy of super with own map prefix
+        public static ItemStack setupNewMap(World worldIn, double worldX, double worldZ, byte scale, boolean trackingPosition, boolean unlimitedTracking) {
```
Comment (mezz) [inline src/test/java/net/minecraftforge/debug/MapDataTest.java]: This is never called

**Review by mezz — CHANGES_REQUESTED** [review]
Tested and it didn't work as expected.







Also please give the map item a name and icon.

**Comment by williewillus** [comment]
Fixed. you should now use the `mapdatatest:empty_custom_map` item to test instead of giving filled map directly.

**Comment by mezz** [comment]
Ok looks like it's working. The test mod's solid red demonstration is not very good but I was able to poke at it and show that it can draw the original map with overlays if it wanted to. Looks good to me.

**Review by mezz — APPROVED** [review]

**Comment by LexManos** [comment]
Again, why can't you just use the normal map files and just stick your data in the correct metadata'd nbt file?

**Comment by williewillus** [comment]
I don't quite get the question.

Vanilla maps are stored in `MapData`, subclass of `WorldSavedData` in the WSD keys `map_<meta>`. The methods on vanilla map items work only like this.

Let's say we need to use a custom subclass of MapData, `CustomMapData`. Ignoring the issue of getting such a subclass synced clientside, the map renderers only call `Items.FILLED_MAP.getMapData` instead of polymorphically calling `ItemMap.getMapData`. As noted above, the vanilla method only queries for the vanilla data. Meaning if I have a meta 0 of my own map, the client will attempt to load and render data from the meta 0 vanilla map because the aforementioned method is not called virtually.

Hope that made some sort of sense.

**Comment by LexManos** [comment]
Yes I understand that it's not called in subclasses what I'm asking is why does it matter? All you're doing is shoving your own data into the map data. From what I see it's all compatible with the vanilla system so literally your map COULD be a vanilla map.

**Comment by williewillus** [comment]
Here's a concrete example with a slightly different rationale: Twilight Forest's Maze Maps.

* They are Y-level aware when mapping
* Thus they need a custom subclass of MapData and thus a custom WSD "mazemap_\<meta\>" to store the y level of creation
  * The y level cannot be stored in the map item itself because all copies of the map need it. One should be able to type `/give @p twilightforest:maze_map 1 <meta>` and continue mapping at the same y level.
* Since they use their own WSD, they don't use the vanilla WSD "map_\<meta\>"
* However, everything clientside only uses the vanilla WSD
* Then, a maze_map of meta 1 will not render the right pixel data, but instead the pixel data of minecraft:filled_map meta 1

Basically the rationale is:
Need custom MapData -> Needs custom WSD name -> client doesn't query my custom WSD name

**Comment by williewillus** [comment]
Another reason for needing custom MapData would be to keep additional metadata with the map (for example, custom decorations #4027)

**Comment by williewillus** [comment]
bump re ^
